From 8f9f0ad9d99e9b0963b6b1aae8a121968da0331d Mon Sep 17 00:00:00 2001
Message-Id: <8f9f0ad9d99e9b0963b6b1aae8a121968da0331d.1464188586.git.jpinto@synopsys.com>
In-Reply-To: <937427def856e507a3e042fac4eb8e4e60fcae0a.1464188586.git.jpinto@synopsys.com>
References: <937427def856e507a3e042fac4eb8e4e60fcae0a.1464188586.git.jpinto@synopsys.com>
From: Joao Pinto <jpinto@synopsys.com>
Date: Wed, 25 May 2016 16:01:02 +0100
Subject: [PATCH 5/5] dwc3 tweak to support dma coherent in ARM64

ARM64 dma is a bit different from the rest of the architectures since kernel
4.4. To enable dma coherent allocation this fix has to be applied.

Signed-off-by: Joao Pinto <jpinto@synopsys.com>
---
 drivers/usb/dwc3/host.c | 13 ++++++++-----
 1 file changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/usb/dwc3/host.c b/drivers/usb/dwc3/host.c
index c679f63..b6ec64b 100644
--- a/drivers/usb/dwc3/host.c
+++ b/drivers/usb/dwc3/host.c
@@ -17,6 +17,7 @@
 
 #include <linux/platform_device.h>
 #include <linux/usb/xhci_pdriver.h>
+#include <linux/of_device.h>
 
 #include "core.h"
 
@@ -32,11 +33,7 @@ int dwc3_host_init(struct dwc3 *dwc)
 		return -ENOMEM;
 	}
 
-	dma_set_coherent_mask(&xhci->dev, dwc->dev->coherent_dma_mask);
-
-	xhci->dev.parent	= dwc->dev;
-	xhci->dev.dma_mask	= dwc->dev->dma_mask;
-	xhci->dev.dma_parms	= dwc->dev->dma_parms;
+	xhci->dev.parent = dwc->dev;
 
 	dwc->xhci = xhci;
 
@@ -62,6 +59,12 @@ int dwc3_host_init(struct dwc3 *dwc)
 	phy_create_lookup(dwc->usb3_generic_phy, "usb3-phy",
 			  dev_name(&xhci->dev));
 
+	if (!dwc->dev->of_node) {
+		xhci->dev.dma_mask = dwc->dev->dma_mask;
+		xhci->dev.dma_parms = dwc->dev->dma_parms;
+	} else 
+		of_dma_configure(&xhci->dev, dwc->dev->of_node);
+
 	ret = platform_device_add(xhci);
 	if (ret) {
 		dev_err(dwc->dev, "failed to register xHCI device\n");
-- 
1.8.1.5

